glib-macros/src/clone.rs

```diff
@@ -0,0 +1,562 @@
+// Copyright 2020, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use crate::utils::crate_ident_new;
+use proc_macro::token_stream::IntoIter as ProcIter;
+use proc_macro::{Delimiter, Group, TokenStream, TokenTree};
+use std::iter::Peekable;
+
+#[derive(Clone, Copy, Debug)]
+enum BorrowKind {
+    Weak,
+    WeakAllowNone,
+    Strong,
+}
+
+impl BorrowKind {
+    fn to_str(self) -> &'static str {
+        match self {
+            Self::Weak => "@weak",
+            Self::WeakAllowNone => "@weak-allow-none",
+            Self::Strong => "@strong",
+        }
+    }
+}
+
+enum WrapperKind {
+    DefaultPanic,
+    DefaultReturn(String),
+}
+
+impl WrapperKind {
+    fn to_str(&self) -> String {
+        match *self {
+            Self::DefaultPanic => "@default-panic".to_owned(),
+            Self::DefaultReturn(ref r) => format!("@default-return {}", r),
+        }
+    }
+
+    fn keyword(&self) -> &'static str {
+        match *self {
+            Self::DefaultPanic => "default-panic",
+            Self::DefaultReturn(_) => "default-return",
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ElemToClone {
+    name: String,
+    alias: Option<String>,
+    borrow_kind: BorrowKind,
+}
+
+impl ElemToClone {
+    fn to_str_before(&self) -> String {
+        match self.borrow_kind {
+            BorrowKind::Weak | BorrowKind::WeakAllowNone => format!(
+                "let {} = {}::clone::Downgrade::downgrade(&{});",
+                if let Some(ref a) = self.alias {
+                    a
+                } else {
+                    &self.name
+                },
+                crate_ident_new(),
+                self.name,
+            ),
+            BorrowKind::Strong => format!(
+                "let {} = {}.clone();",
+                if let Some(ref a) = self.alias {
+                    a
+                } else {
+                    &self.name
+                },
+                self.name,
+            ),
+        }
+    }
+
+    fn to_str_after(&self, wrapper_kind: &WrapperKind) -> String {
+        match (self.borrow_kind, wrapper_kind) {
+            (BorrowKind::Weak, WrapperKind::DefaultPanic) => {
+                let name = if let Some(ref a) = self.alias {
+                    a
+                } else {
+                    &self.name
+                };
+                format!(
+                    "\
+let {0} = match {1}::clone::Upgrade::upgrade(&{0}) {{
```

Comment:
We use syn and quote in this crate elsewhere, can you check if using them makes things simpler / less verbose here? Also CC @gdesmott, who probably has an opinion on that.

Reply:
I'm far from being a macro expert but `quote` and `syn` makes things much easier and readable from my (small) experience. I'd advise to keep using them.

Reply:
Using syn makes no sense here as the macro does not really contain Rust but only something based on Rust tokens, and the actual closure block we take as-is.

quote might make sense and I'll look at that next week, together with improving the parsing a bit. But for now this is all ok, I don't think there are bugs :)